src/BaseEnum.php

```diff
@@ -2,40 +2,68 @@
 
 namespace Greg0ire\Enum;
 
-abstract class BaseEnum
+class BaseEnum extends AbstractBaseEnum
 {
-    private static $constCache = array();
+    private static $classes;
+
+    public function __construct($className = null)
+    {
+        if ($className) {
+            $this->addClass($className);
+        }
+    }
 
     /**
-     * Uses reflection to find the constants defined in the class and cache
-     * them in a local property for performance, before returning them.
+     * Add new class|interface name will be used
+     * to find the constants defined
+     *
+     * @param string|array $className the name of the class or interface
+     */
+    public function addClass($className)
```

Comment:
Why have two separate classes ? Why not have an overridable static function `getEnumTypes()` that by default, would return the current class ? This would maintain BC…